Improve doc and implementation of OakUnsafeDirectBuffer.getByteBuffer

dynaxis opened the pull request:
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

---

This PR improves doc comment on `OakUnsafeDirectBuffer.getByteBuffer` to better reflect the current implementation after removing internal uses of `ByteBuffer`s.

Also it reimplements proper protection of read-only Oak buffers by returning read only view of `ByteBuffer` if source buffer implementation is read only. Such a measure was mistakenly removed in the course of the recent elimination of all `ByteBuffer` usage.

sanastas reviewed (commented):
The suggested changes looks good. Thank you!

Can you please a test to check that it is impossible to write into the read-only OakUnsafeDirectBuffer? It is relatively easy, for example in `InternalOakMapTest` there is 

```
  private static Integer slowDeserialize(OakScopedReadBuffer bb) {
        try {
            Thread.sleep(LONG_TRANSFORMATION_DELAY);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return OakCommonBuildersFactory.DEFAULT_INT_SERIALIZER.deserialize(bb);
    }
```

so it is possible to cast `OakScopedReadBuffer bb` and to try to write into it... What do you think?

dynaxis commented:
Makes sense. I'll update the current PR soon.

dynaxis commented:
I added tests for all the implementations of `OakUnsafeDirectBuffer` to test writability of the returned `ByteBuffer`. And found that `UnscopeBuffer` and its subclass violate the spec and also fixed it.

In the end, the tests are the most complex part of this PR. :)

liran-funaro commented on core/src/test/java/com/yahoo/oak/KeyBufferTest.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @Test
+    public void byteBufferShouldBeReadOnly() {
```
Can you merge all the test classes you added into one test class? They seem to fit the same pattern and they test similar things. It makes sense to group them together.

dynaxis replied:
@liran-funaro The reason I separated them that way was that it's a test among many others that may be added in the future.  If you still think it's better to consolidate them all, then I'll add another commit for doing so. Please give your final answer here. Thanks.

liran-funaro replied:
Thanks for the explanation. Since we don't have other tests for these buffers yet, I think it is better to consolidate them into a single class. But I do agree each test should be in its own method.

liran-funaro replied:
Great. Thank you for your PR.

dynaxis replied:
I pushed another commit to fulfill the request from you. Thank you.

liran-funaro reviewed (changes requested):
The PR looks good. I request only a minor change in the tests (see below).

dynaxis commented:
I haven't run all of the tests except for those added by me due to some other tests failing in my environment. So do you think the failing test is due to this PR?

sanastas commented:
The failures looks unrelated to your PR and also to Oak in general, as this is just a way of string creation. More than that, all the tests are running upon each of your commits/pushes and you have "All checks have passed" meaning there were no failures on mvn flow.

Is it the first time you run all the tests? Maybe it is related to your local environment, Java version, etc.

sanastas commented:
@dynaxis , pleasure to work with you! Your PR looks very good! Thank you! I am merging to master.

dynaxis commented:
Happy to work with you and to contribute a small bit to the project I rely on. Thank you.
